src/runtime/tuple_parser.c

```diff
@@ -185,34 +215,43 @@ closure_function(4, 1, parser, name_complete,
     heap h = bound(h);
     completion vc = closure(h, value_complete, bound(t), intern(b), bound(check));
     // not sure why we have to violate typing
-    parser pv = (void *)closure(h, parse_value, h, vc, bound(err));
-    return ignore_whitespace(h, (void *)closure(h, dispatch_property, h, pv, bound(err)));
+    parser pv = (parser)closure(h, parse_value, h, vc, bound(err));
+    parser p = ignore_whitespace(h, (parser)closure(h, dispatch_property, h, pv, bound(err)));
+    closure_finish();
+    return p;
 }
 
 closure_function(3, 1, parser, parse_name,
                  heap, h, completion, c, buffer, b,
                  character, in)
 {
     heap h = bound(h);
+    parser p;
     if (in == '"') {
-        return combinate(h, closure(h, quoted_string, h, bound(c), bound(b)));
+        p = combinate(h, closure(h, quoted_string, h, bound(c), bound(b)));
+    } else {
+        parser q = combinate(h, closure(h, terminal, bound(c), name_terminal, bound(b)));
+        p = apply(q, in);
     }
-
-    parser p = combinate(h, closure(h, terminal, bound(c), name_terminal, bound(b)));
-    return apply(p, in);
+    closure_finish();
+    return p;
 }
 
 static parser is_end_of_tuple(struct _closure_is_end_of_tuple *__self, parser self, character in)
 {
     heap h = bound(h);
     if (in == ')') {
-        return apply(bound(c), bound(t));
+        parser p = apply(bound(c), bound(t));
+        deallocate_closure(self);
+        closure_finish();
+        return p;
     }
 
+    // XXX
```

Comment:
what is the xx here ? (closure finish above)

Reply:
just cruft placed on first pass while I was trying to understand the code...no issue there that I see now